File: sdks/elixir/lib/bcs.ex
defmodule Bcs do
  import Bitwise

  @moduledoc """
  Binary Canonical Serialization (BCS) for Elixir.

  BCS is a deterministic binary serialization format that guarantees
  canonical representation - every value has exactly one valid encoding.

  This library provides two APIs:

  ## Manual API

  Explicit control over serialization using `Bcs.Serializer` and `Bcs.Deserializer`:

      alias Bcs.{Serializer, Deserializer}

      # Serialization
      data =
        Serializer.new()
        |> Serializer.write_u8(1)
        |> Serializer.write_u64(100)
        |> Serializer.write_string("hello")
        |> Serializer.to_bytes()

      # Deserialization
      {value1, rest} = Deserializer.read_u8!(data)
      {value2, rest} = Deserializer.read_u64!(rest)
      {value3, rest} = Deserializer.read_string!(rest)
      Deserializer.check_end!(rest)

  ## Idiomatic API (coming soon)

  Using structs with the `Bcs.Struct` DSL:

      defmodule Transfer do
        use Bcs.Struct

        bcs_struct do
          field :sender, {:fixed_bytes, 32}
          field :recipient, {:fixed_bytes, 32}
          field :amount, :u64
        end
      end

      transfer = %Transfer{sender: addr1, recipient: addr2, amount: 1000}
      data = Bcs.serialize(transfer)
      {:ok, transfer2} = Bcs.deserialize(Transfer, data)

  ## Supported Types

  - Booleans: `bool`
  - Unsigned integers: `u8`, `u16`, `u32`, `u64`, `u128`, `u256`
  - Signed integers: `i8`, `i16`, `i32`, `i64`, `i128`, `i256`
  - Byte arrays: `bytes` (length-prefixed)
  - Strings: `string` (UTF-8, length-prefixed)
  - Fixed bytes: `{:fixed_bytes, length}`
  - Options: `{:option, inner_type}`
  - Vectors: `{:vector, element_type}`
  - Maps: `{:map, key_type, value_type}`
  - Enums: via variant index

  ## Constants

  - `MAX_SEQUENCE_LENGTH`: 2^31 - 1 (2,147,483,647)
  - `MAX_CONTAINER_DEPTH`: 500

  """

  @max_sequence_length (1 <<< 31) - 1
  @max_container_depth 500

  @doc """
  Maximum allowed sequence length.
  """
  @spec max_sequence_length() :: non_neg_integer()
  def max_sequence_length, do: @max_sequence_length

  @doc """
  Maximum allowed container nesting depth.
  """
  @spec max_container_depth() :: non_neg_integer()
  def max_container_depth, do: @max_container_depth

  # Re-export commonly used modules
  defdelegate serialize_u8(value), to: __MODULE__.Convenience
  defdelegate serialize_u64(value), to: __MODULE__.Convenience
  defdelegate serialize_string(value), to: __MODULE__.Convenience
  defdelegate serialize_bytes(value), to: __MODULE__.Convenience

  defdelegate deserialize_u8(data), to: __MODULE__.Convenience
  defdelegate deserialize_u64(data), to: __MODULE__.Convenience
  defdelegate deserialize_string(data), to: __MODULE__.Convenience
  defdelegate deserialize_bytes(data), to: __MODULE__.Convenience
end

defmodule Bcs.Convenience do
  @moduledoc false
  # Convenience functions for common operations

  alias Bcs.{Deserializer, Serializer}

  @doc """
  Serialize a u8 value.
  """
  @spec serialize_u8(non_neg_integer()) :: binary()
  def serialize_u8(value) do
    Serializer.new()
    |> Serializer.write_u8(value)
    |> Serializer.to_bytes()
  end

  @doc """
  Serialize a u64 value.
  """
  @spec serialize_u64(non_neg_integer()) :: binary()
  def serialize_u64(value) do
    Serializer.new()
    |> Serializer.write_u64(value)
    |> Serializer.to_bytes()
  end

  @doc """
  Serialize a string value.
  """
  @spec serialize_string(String.t()) :: binary()
  def serialize_string(value) do
    Serializer.new()
    |> Serializer.write_string(value)
    |> Serializer.to_bytes()
  end

  @doc """
  Serialize a bytes value.
  """
  @spec serialize_bytes(binary()) :: binary()
  def serialize_bytes(value) do
    Serializer.new()
    |> Serializer.write_bytes(value)
    |> Serializer.to_bytes()
  end

  @doc """
  Deserialize a u8 value.
  """
  @spec deserialize_u8(binary()) :: {:ok, non_neg_integer()} | {:error, Bcs.Error.t()}
  def deserialize_u8(data) do
    with {:ok, {value, rest}} <- Deserializer.read_u8(data),
         :ok <- Deserializer.check_end(rest) do
      {:ok, value}
    end
  end

  @doc """
  Deserialize a u64 value.
  """
  @spec deserialize_u64(binary()) :: {:ok, non_neg_integer()} | {:error, Bcs.Error.t()}
  def deserialize_u64(data) do
    with {:ok, {value, rest}} <- Deserializer.read_u64(data),
         :ok <- Deserializer.check_end(rest) do
      {:ok, value}
    end
  end

  @doc """
  Deserialize a string value.
  """
  @spec deserialize_string(binary()) :: {:ok, String.t()} | {:error, Bcs.Error.t()}
  def deserialize_string(data) do
    with {:ok, {value, rest}} <- Deserializer.read_string(data),
         :ok <- Deserializer.check_end(rest) do
      {:ok, value}
    end
  end

  @doc """
  Deserialize a bytes value.
  """
  @spec deserialize_bytes(binary()) :: {:ok, binary()} | {:error, Bcs.Error.t()}
  def deserialize_bytes(data) do
    with {:ok, {value, rest}} <- Deserializer.read_bytes(data),
         :ok <- Deserializer.check_end(rest) do
      {:ok, value}
    end
  end
end

File: sdks/ruby/lib/bcs.rb
# frozen_string_literal: true

# Copyright (c) BCS SDK Contributors
# SPDX-License-Identifier: Apache-2.0

require_relative "bcs/version"
require_relative "bcs/constants"
require_relative "bcs/errors"
require_relative "bcs/uleb128"
require_relative "bcs/serializer"
require_relative "bcs/deserializer"

# Binary Canonical Serialization (BCS) - Ruby Implementation
#
# BCS is a deterministic binary serialization format designed for
# canonical representation of data structures.
#
# @example Serialize and deserialize
#   # Serialize
#   ser = BCS::Serializer.new
#   ser.write_u64(12345)
#   ser.write_string("hello")
#   ser.write_bool(true)
#   bytes = ser.to_bytes
#
#   # Deserialize
#   des = BCS::Deserializer.new(bytes)
#   num = des.read_u64
#   str = des.read_string
#   flag = des.read_bool
#   des.check_end
#
# @example Using object pooling for high-throughput scenarios
#   # Acquire from pool
#   ser = BCS.acquire_serializer
#   ser.write_u64(12345)
#   bytes = ser.to_bytes
#   BCS.release_serializer(ser)  # Return to pool
#
module BCS
  # Simple object pool for Serializers (thread-safe via Mutex)
  @serializer_pool = []
  @serializer_pool_mutex = Mutex.new
  @serializer_pool_max_size = 16

  class << self
    # Acquire a Serializer from the pool (or create new if pool is empty)
    # @param capacity [Integer] Initial buffer capacity
    # @return [Serializer] A serializer instance
    def acquire_serializer(capacity: Serializer::DEFAULT_CAPACITY)
      @serializer_pool_mutex.synchronize do
        ser = @serializer_pool.pop
        return ser if ser

        Serializer.new(capacity: capacity)
      end
    end

    # Release a Serializer back to the pool for reuse
    # @param serializer [Serializer] The serializer to release
    # @return [void]
    def release_serializer(serializer)
      serializer.clear
      @serializer_pool_mutex.synchronize do
        @serializer_pool << serializer if @serializer_pool.size < @serializer_pool_max_size
      end
    end

    # Use a pooled serializer with a block (automatically released)
    # @param capacity [Integer] Initial buffer capacity
    # @yield [Serializer] The serializer to use
    # @return [Object] The return value of the block
    def with_serializer(capacity: Serializer::DEFAULT_CAPACITY)
      ser = acquire_serializer(capacity: capacity)
      begin
        yield ser
      ensure
        release_serializer(ser)
      end
    end
    # Convenience methods for single-value serialization

    def serialize_u8(value)
      Serializer.new.write_u8(value).to_bytes
    end

    def serialize_u16(value)
      Serializer.new.write_u16(value).to_bytes
    end

    def serialize_u32(value)
      Serializer.new.write_u32(value).to_bytes
    end

    def serialize_u64(value)
      Serializer.new.write_u64(value).to_bytes
    end

    def serialize_bool(value)
      Serializer.new.write_bool(value).to_bytes
    end

    def serialize_string(value)
      Serializer.new.write_string(value).to_bytes
    end

    def serialize_bytes(value)
      Serializer.new.write_bytes(value).to_bytes
    end

    # Convenience methods for single-value deserialization

    def deserialize_u8(data)
      des = Deserializer.new(data)
      value = des.read_u8
      des.check_end
      value
    end

    def deserialize_u16(data)
      des = Deserializer.new(data)
      value = des.read_u16
      des.check_end
      value
    end

    def deserialize_u32(data)
      des = Deserializer.new(data)
      value = des.read_u32
      des.check_end
      value
    end

    def deserialize_u64(data)
      des = Deserializer.new(data)
      value = des.read_u64
      des.check_end
      value
    end

    def deserialize_bool(data)
      des = Deserializer.new(data)
      value = des.read_bool
      des.check_end
      value
    end

    def deserialize_string(data)
      des = Deserializer.new(data)
      value = des.read_string
      des.check_end
      value
    end

    def deserialize_bytes(data)
      des = Deserializer.new(data)
      value = des.read_bytes
      des.check_end
      value
    end

    # Hex utilities

    def bytes_to_hex(bytes)
      bytes.pack("C*").unpack1("H*")
    end

    def hex_to_bytes(hex)
      [hex].pack("H*").unpack("C*")
    end
  end
end
